lib/ansible/module_utils/urls.py

```diff
@@ -481,31 +483,57 @@ def redirect_request(self, req, fp, code, msg, hdrs, newurl):
             elif follow_redirects in ['no', 'none', False]:
                 raise urllib_error.HTTPError(newurl, code, msg, hdrs, fp)
 
+            method = req.get_method()
+
             do_redirect = False
             if follow_redirects in ['all', 'yes', True]:
-                do_redirect = (code >= 300 and code < 400)
+                if 300 > code or code >= 400:
+                    raise urllib_error.HTTPError(req.get_full_url(), code, msg, hdrs, fp)
 
-            elif follow_redirects == 'safe':
-                m = req.get_method()
-                do_redirect = (code >= 300 and code < 400 and m in ('GET', 'HEAD'))
+            if follow_redirects == 'safe':
+                if 300 > code or code >= 400 or method not in ('GET', 'HEAD'):
+                    raise urllib_error.HTTPError(req.get_full_url(), code, msg, hdrs, fp)
 
-            if do_redirect:
-                # be conciliant with URIs containing a space
-                newurl = newurl.replace(' ', '%20')
-                newheaders = dict((k, v) for k, v in req.headers.items()
-                                  if k.lower() not in ("content-length", "content-type"))
-                try:
-                    # Python 2-3.3
-                    origin_req_host = req.get_origin_req_host()
-                except AttributeError:
-                    # Python 3.4+
-                    origin_req_host = req.origin_req_host
-                return urllib_request.Request(newurl,
-                                              headers=newheaders,
-                                              origin_req_host=origin_req_host,
-                                              unverifiable=True)
+            try:
+                # Python 2-3.3
+                data = req.get_data()
+                origin_req_host = req.get_origin_req_host()
+            except AttributeError:
+                # Python 3.4+
+                data = req.data
+                origin_req_host = req.origin_req_host
+
+            # Be conciliant with URIs containing a space
+            newurl = newurl.replace(' ', '%20')
+
+            # Suport redirect with payload and original headers
+            if code in (307, 308):
+                headers = req.headers
             else:
-                raise urllib_error.HTTPError(req.get_full_url(), code, msg, hdrs, fp)
```

Comment:
The removal of this final raising of `HTTPError` means that there are some scenarios that can succeed, such as when given invalid redirection options.

```
RedirectHandlerFactory(follow_redirects='invalid', validate_certs=True)
```

Previous behavior was to raise an `HTTPError` in this situation, somewhat similar to `follow_redirects='none'`

Reply:
@sivel I fixed this in the last commit